Update Interactable property code to utilize shader int keys instead of strings

Troy-Ferrell opened the pull request:
## Overview

InteractableTheme classes are able to update shader properties such as color or glow, etc in response to scene interactions such as focus or click. These theme handler classes updates shaders in Unity via string properties. Whenever a property is updated via a string key such as GetFloat(), Unity will automatically call Shader.PropertyToId() to get it's int key and then perform the get or the set. 

This change caches the integer property keys for all property strings in a given theme class so as to help with performance when calling the get/set functions at runtime

## Changes
- Fixes: #4228 
Optimize Shader property and Material access on Interactable*Theme class

## Verification
Confirmed HandInteraction scene worked with Interactable classes on changes such as focus or clicks, etc.

wiwei commented:
/azp run mrtk_pr

wiwei commented:
/azp run mrtk_docs

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Themes/InteractableThemeProperty.cs:
```diff
@@ -16,20 +17,43 @@ public class InteractableThemeProperty
+            if (ShaderPropertyIDs == null)
+            {
+                ShaderPropertyIDs = new List<int>(ShaderOptionNames.Count);
+                for(int i = 0; i < this.ShaderOptionNames.Count; i++)
```
```suggestion
                for (int i = 0; i < this.ShaderOptionNames.Count; i++)
```

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Themes/InteractableThemeProperty.cs:
```diff
@@ -16,20 +17,43 @@ public class InteractableThemeProperty
         public InteractableThemePropertyValue Default;
         public string ShaderName;
 
-        public string GetShaderPropId()
```
@killerantz do you think that the chances are fairly low that someone else would be using this outside of the interactable family? This is technically a breaking change (in this case if we have any thoughts of there being usage we might keep this around and just mark it obsolete - or just redirect it to the renamed thing).

Assuming it's fairly low it seems reasonable to do things this way, just wanted to confirm

killerantz replied:
As far as I know they are only used in Interactable and this is pretty deep. Just make sure it's documented well, in the off chance someone is using it, they can easily update their code to store an int instead of string.

Troy-Ferrell replied:
Added documentation in code and will add documentation to github page

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Scripts/Themes/InteractableThemeProperty.cs:
```diff
@@ -16,20 +17,43 @@ public class InteractableThemeProperty
+        private List<int> ShaderPropertyIDs = null;
+        private const string DefaultProperty = "_Color";
+
+        public int GetShaderPropertyId()
```
Nit: Can you also add doccomments to these functions (i.e. maybe recommending the int flavor over the string flavor?).

Troy-Ferrell replied:
Documentation comments added

wiwei commented:
Overall lgtm, just a few minor q's

Troy-Ferrell commented:
@Cameron-Micka , @wiwei , @killerantz , thanks! 

I did run the HandInteraction examples scene and validated all interactables changed properties as expected both in editor & on device, particularly the piano keys were easy to validate. I haven't found the InteractionProperties used anywhere else except Interactable (both runtime & inspectors). I know it is a breaking change but as already stated, minor and unlikely someone will be using it directly as well as it is an easy thing to fix if indeed the case. 

I will definitely be sure to updated documentation

Cameron-Micka commented:
@troy

> @Cameron-Micka , @wiwei , @killerantz , thanks!
> I did run the HandInteraction examples scene and validated all interactables changed properties as expected both in editor & on device, particularly the piano keys were easy to validate. I haven't found the InteractionProperties used anywhere else except Interactable (both runtime & inspectors). I know it is a breaking change but as already stated, minor and unlikely someone will be using it directly as well as it is an easy thing to fix if indeed the case.
> I will definitely be sure to updated documentation

Cameron-Micka commented:
> @Cameron-Micka , @wiwei , @killerantz , thanks!
> I did run the HandInteraction examples scene and validated all interactables changed properties as expected both in editor & on device, particularly the piano keys were easy to validate. I haven't found the InteractionProperties used anywhere else except Interactable (both runtime & inspectors). I know it is a breaking change but as already stated, minor and unlikely someone will be using it directly as well as it is an easy thing to fix if indeed the case.
> I will definitely be sure to updated documentation

That sounds great! If you are looking to test more, @killerantz has a scene called "InteractablesExamples"

luis-valverde-ms commented:
Hey @Troy-Ferrell , any update on this? Anything I can do to help?

Troy-Ferrell commented:
> Hey @Troy-Ferrell , any update on this? Anything I can do to help?

Sorry for the delay. I got caught up in work and travel recently. 

1) Fixed space in for loop
2) Tested "InteractablesExamples" scene and objects update correctly on focus, etc
3) Added documentation to code explaining preferrence of integer key over string key
4) Updating this page to reflect this point as well:
https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_Interactable.html

luis-valverde-ms commented:
@Troy-Ferrell , thanks for the push. Ping @wiwei if/when this is ready for check in.

Troy-Ferrell commented:
@wiwei can you merge this in? Thanks!

wiwei commented:
Yes, sorry for the delay!

wiwei commented:
/azp run mrtk_pr

wiwei commented:
/azp run mrtk_docs

keveleigh commented:
Merging with two approvals and successful builds.
